composeApp/src/desktopMain/kotlin/org/ooni/probe/data/DeepLinkHandler.kt

```diff
@@ -0,0 +1,278 @@
+package org.ooni.probe.data
+
+import co.touchlab.kermit.Logger
+import org.ooni.probe.APP_ID
+import tk.pratanumandal.unique4j.Unique4j
+import tk.pratanumandal.unique4j.exception.Unique4jException
+import java.io.File
+import java.nio.file.Files
+import java.nio.file.Paths
+import java.util.Date
+
+class DeepLinkHandler {
+    private var unique: Unique4j? = null
+    private val messageListeners = mutableListOf<(String?) -> Unit>()
+
+    fun addMessageListener(listener: (String?) -> Unit) {
+        messageListeners.add(listener)
+    }
+
+    fun parseDeepLink(deepLink: String): Map<String, String> {
+        val params = mutableMapOf<String, String>()
+
+        if (deepLink.startsWith("ooni://")) {
+            val parts = deepLink.removePrefix("ooni://").split("/")
+            if (parts.size >= 2 && parts[0] == "runv2") {
+                params["command"] = "runv2"
+                params["testId"] = parts[1]
+            }
+        }
+
+        return params
+    }
+
+    fun initialize(args: Array<String>) {
+        try {
+            unique = object : Unique4j(APP_ID) {
+                override fun receiveMessage(message: String) {
+                    Logger.d("Received message: $message")
+
+                    // Parse deep link if it starts with ooni://
+                    if (message.startsWith("ooni://")) {
+                        val params = parseDeepLink(message)
+                        handleDeepLinkCommand(message, params)
+                    }
+                }
+
+                override fun sendMessage(): String {
+                    // Check for deep links in arguments
+                    val deepLink = findDeepLinkInArgs(args)
+                    return deepLink ?: "Application launched without deep link at ${Date()}"
+                }
+
+                override fun handleException(exception: Exception) {
+                    Logger.e(exception) { "Exception occurred" }
+                }
+
+                override fun beforeExit() {
+                    Logger.d("Exiting subsequent instance.")
+                }
+            }
+
+            // Try to acquire lock
+            val lockFlag = unique!!.acquireLock()
+            if (lockFlag) {
+                Logger.d("Application instance started successfully")
+
+                // If this is the first instance and has a deep link, handle it directly
+                val deepLink = findDeepLinkInArgs(args)
+                if (deepLink != null) {
+                    val params = parseDeepLink(deepLink)
+                    // messageListeners.forEach { it("Initial deep link: $deepLink, params: $params") }
+                    handleDeepLinkCommand(deepLink, params)
+                }
+
+                // TODO: Implement with conveyor
+                // Register as a protocol handler if not already registered
+                registerProtocolHandler()
+            } else {
+                Logger.d("Could not acquire lock - this should not happen")
```

Comment:
Isn't this what happens with the second instance? And should we ignore the args from the second instance? Would a deeplink work for an already existing app?

Reply:
The deeplink works if there is an instance already. 

The lock is for unique4j to know the status